_posts/2021-05-18-My-First-Steps-in-OpenSearch-Plugins.markdown

```diff
@@ -0,0 +1,618 @@
+  - [Installing the plugin](#installing-the-plugin)
+  - [Running OpenSearch with our new plugin](#running-opensearch-with-our-new-plugin)
+  - [Uninstalling the plugin](#uninstalling-the-plugin)
+- [Key takeaways from creating my first plugin](#key-takeaways-from-creating-my-first-plugin)
+
+## What are OpenSearch Plugins? 
+
+Software plugins are, in essence, stand alone components __add features and capabilities__ to an existing core program. 
+
+![Figure 1: Plugins architectural diagram]({{ site.baseurl }}/assets/media/blog-images/2021-05-18-My-First-Steps-in-OpenSearch-Plugins/plugins-architectural-diagram.jpg){: .img-fluid }
+
+Normally, when viewing a plugin's code you wouldn't see *any* point of entry (no `main()`). That's because the plugin's points of entry are the interface functions that it implements. 
+
+OpenSearch plugins are no different.
+
+As an analogy, think of how you can replace lenses and flash elements on a camera body. The core project is the part that _takes the picture_. It contains things a camera must have such as a light sensor and the software required to record the sensor's information. Plugins then would be what you can _connect to the camera body_. 
+
+Take for example, a wide angle lens to take pictures of landscapes. Maybe a cool feature would be to keep the camera still without touching it? Or taking it under water? You could install a tripod or a protective waterproof case respectively. Consider also that these plugins connect to the camera body in some way. To create one yourself you must first know 1) how to connect to the camera, and 2) how to interact with it via these connections -- this is called the **plugin interface**.
+
+If you are designing a way for the camera to store more images on a new form of memory chip, then you don't need to know anything about the interface of the lenses. However, you would have to know a great deal about how this camera sends/retrieves its recorded file via the memory chip connectors. 
+
+In other words, the plugin interface is not necessarily the same for every plugin.
+
+
+The beauty of this architecture is that it enables you to customize the core program to your needs, without having to contribute to the core project directly (Imagine having to buy a new camera for every different lens instead of switching the lenses!).
+
+Let's elaborate on the two different kinds of OpenSearch plugins:
+
+### 1. Core Plugins
+Some plugins are included and maintained in the core project, called __Core Plugins__. 
+
+An example of one of these is the [`repository-s3 plugin`](https://github.com/opensearch-project/OpenSearch/tree/main/plugins/repository-s3/src) that implements snapshot and restore capabilities using AWS's S3 buckets as the repository. 
+
+The importance of the Core Plugins is that they are considered essential *features* for the community and therefore maintained as part of the project.  Which means they are *in the open source project repository*.
```

Comment:
I dont know if we really call them `Core Plugins`. As dblock@ said, @nknize is probably the one who can categorize this.

Reply:
Thanks! Do you suggest I ignore the "plugin types" for the sake of this blog post?

Reply:
This will require some more editing - there are basically three places plugins may be: 
- the opensearch project 
- outside the project (still open-source though) 
- in the opensearch repo (like the s3-repository plugin) 

Perhaps this is not at all important to this blog post and I should reframe this section to elaborate on the benifits of the community involvement in creating plugins.
WDYT?

Reply:
Yeah I like that, let's keep it generic and explain where they live rather than having a keyword for them. 

I think it's important to have this section as this is the starting point for thinking of "Where should my new plugin live?"
I like the way you described:
1. OpenSearch repo - Only really essential plugins like s3 repository sit there.
2. OpenSearch-project - All the plugins which are officially supported by the project.
3. Community plugins - Any where in this world, could be open/closed source.

Reply:
@saratvemulapalli I edited it to reflect what @dblock was saying in his comments on this very same issue